Print a code snippet with the panicking expression

estebank opened the pull request:
When a backtrace is being printed out, and the source code file where a panic originated is available, print a snippet pointing at the expression that panicked.

```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1'
 --> tests/ui/panics/singleline-snippet.rs:3:13
1 | fn main() {
2 |     let x = vec![vec![vec![vec![1,2,3]]]];
3 |     let _ = x[0][0][1][5];
  |             ^^^^^^^^^^
4 | }
stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at ./library/core/src/panicking.rs:64:14
   2: core::panicking::panic_bounds_check
             at ./library/core/src/panicking.rs:159:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
   6: singleline_snippet::main
   7: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1'
 --> tests/ui/panics/multiline-2.rs:3:13
1 | fn main() {
2 |     let x = vec![vec![vec![vec![1,2,3]]]];
3 |     let _ = x[0]
  | ____________^
4 ||        [0][1][5];
  ||_____________^
5 | }
stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at ./library/core/src/panicking.rs:64:14
   2: core::panicking::panic_bounds_check
             at ./library/core/src/panicking.rs:159:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
   6: multiline_2::main
   7: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

estebank committed:
Print a code snippet with the panicking expression

When a backtrace is being printed out, and the source code file where a
panic originated is available, print a snippet pointing at the
expression that panicked.

```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1'
 --> tests/ui/panics/singleline-snippet.rs:3:13
1 | fn main() {
2 |     let x = vec![vec![vec![vec![1,2,3]]]];
3 |     let _ = x[0][0][1][5];
  |             ^^^^^^^^^^
4 | }
stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at ./library/core/src/panicking.rs:64:14
   2: core::panicking::panic_bounds_check
             at ./library/core/src/panicking.rs:159:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
   6: singleline_snippet::main
   7: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1'
 --> tests/ui/panics/multiline-2.rs:3:13
1 | fn main() {
2 |     let x = vec![vec![vec![vec![1,2,3]]]];
3 |     let _ = x[0]
  | ____________^
4 ||        [0][1][5];
  ||_____________^
5 | }
stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at ./library/core/src/panicking.rs:64:14
   2: core::panicking::panic_bounds_check
             at ./library/core/src/panicking.rs:159:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
   6: multiline_2::main
   7: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

estebank commented:
For some reason, adding tests for these (even restricted to `--stage 2`) didn't actually trigger the new code. I'm slightly baffled.

estebank commented:
https://user-images.githubusercontent.com/1606434/219252559-e61f8baa-9504-4c7c-afde-528cc8151254.mov

estebank commented:
We can start reviewing and cleaning up, but lets not merge until https://github.com/rust-lang/compiler-team/issues/591 is approved.

bjorn3 commented:
Would it be possible to include a file hash to prevent confusing messages if the source file was changed in the mean time? Maybe the hash could be added right before the filename string the Location refers too? I think that would significantly reduce the size impact compared to adding it to Location directly.

joboet commented:
> Would it be possible to include a file hash to prevent confusing messages if the source file was changed in the mean time? Maybe the hash could be added right before the filename string the Location refers too? I think that would significantly reduce the size impact compared to adding it to Location directly.

I think something like that is very much necessary here, otherwise this can escalate security vulnerabilities in binaries with known bugs: If a high-privilege process can be crashed from a low-privilege position, it can be used to print out (parts of) files requiring higher privilege levels by using symbolic links. 

For instance, consider something like a `rustc`-daemon running from `root` (a stupid thing to do, but bear with me). Reading out the secret file `/secret/secret` normally requires root, but by placing a symbolic link in a folder like `/dist/src/compiler/[path to source with bug]` linking to `/secret/secret` and getting the compiler to ICE at the right place, the panic message prints out (parts of) the secret file.

estebank commented:
@joboet would your concerns be alleviated if this was an explicitly opt-in behavior, needing `RUST_BACKTRACE=short,snippet`, for example? The idea of a hash would be useful to prevent from accidentally showing a file that has since been modified, but I worry that it doesn't raise the bar enough to prevent the case you're highlighting enough. I was operating under the assumption that for that vector to be useful the attacker would have to modify the binary and convince you to run it, at which point you have bigger problems than this, but that's because I wasn't aware [you could symlink to a directory you don't have permissions for](https://unix.stackexchange.com/questions/358690/can-i-create-a-symbolic-link-to-a-file-i-cant-access).

bjorn3 commented:
Would it work to force snippets to be disabled under the same conditions that the dynamic linker ignores LD_PRELOAD and to forbid resolving symlinks while loading the snippet file using O_NOFOLLOW?

joboet commented:
> @joboet would your concerns be alleviated if this was an explicitly opt-in behavior, needing `RUST_BACKTRACE=short,snippet`, for example?

Yes, I think that would work very well. If I remember correctly, the environment variables of a process can only be controlled when at the same privilege level.

> The idea of a hash would be useful to prevent from accidentally showing a file that has since been modified, but I worry that it doesn't raise the bar enough to prevent the case you're highlighting enough. 

It would have to be a cryptographic hash, but I think that should be enough to close the attack vector.

> [...] and to forbid resolving symlinks while loading the snippet file using O_NOFOLLOW?

Are the paths printed in `Location` canonicalized? If so, this would make sense to be. Otherwise that is a bit too restrictive IMHO, since it prevents snippets from being printed if some component in path to the actual source directory is a symlink, which might be hard to explain to Rust users running into that situation.

bjorn3 commented:
> Yes, I think that would work very well. If I remember correctly, the environment variables of a process can only be controlled when at the same privilege level.

setuid programs inherit all env vars, hence why LD_PRELOAD needs to be ignored for them.

> Otherwise that is a bit too restrictive IMHO, since it prevents snippets from being printed if some component in path to the actual source directory is a symlink, which might be hard to explain to Rust users running into that situation.

A message like "not following symlink for code snippet for security reasons" may work, right?

bjorn3 commented:
> It would have to be a cryptographic hash, but I think that should be enough to close the attack vector.

Rustc uses MD5 for source hashing because not all DWARF consumers support anything better. And the better thing is SHA1 I believe. And even if it was cryptographically secure, it would still leak if the file on the disk exactly matches the source file.

estebank commented:
> A message like "not following symlink for code snippet for security reasons" may work, right?

Checking for symlinks would be trivial to add.

> Rustc uses MD5 for source hashing because not all DWARF consumers support anything better.

Even as weak as that is, it would help for the case of "modified file.rs after executing `cargo run` and *then* a panic happened".

> it would still leak if the file on the disk exactly matches the source file.

What is the vector here? You have access to stderr (or the logs), then what is the avenue to do something they can't already do? If anything, this would be a tool for people in control of their services but poor deployment management to figure out if the deployed version matches a certain version of the source code 😅

estebank committed:
Check if the path to the panicking file is a symlink

bjorn3 commented:
> Even as weak as that is, it would help for the case of "modified file.rs after executing cargo run and then a panic happened".

Indeed. That is exactly what I suggested it for. I just don't like it being used for security.

> What is the vector here? You have access to stderr (or the logs), then what is the avenue to do something they can't already do? If anything, this would be a tool for people in control of their services but poor deployment management to figure out if the deployed version matches a certain version of the source code 😅

If a source file embeds a secret you may be able to slowly leak the secret. I get that it is a bit far fetched though.

bjorn3 commented on library/std/src/panicking.rs:
```diff
@@ -261,7 +262,11 @@ fn default_hook(info: &PanicInfo<'_>) {
+            // We check if `path` is a symlink to avoid people creating a symlink for a file they
+            // don't have permissions, but that the running binary does, to use it as an
+            // exfiltration vector.
+            if !path.is_symlink() && let Ok(file) = File::open(path) {
```
This doesn't help if one of the directories in the path is a symlink. You need O_NOFOLLOW to handle directories without a TOCTOU bug.

estebank replied:
Changed. We'd need something equivalent for Windows as well.

estebank replied:
Ended up pulling back from this and instead walked up the path and checked every one for being a symlink with `is_symlink` because that will work in all platforms.

WaffleLapkin replied:
This returned the TOCTOU bug though, didn't it? The change may happen between the check and `File::open` call.

estebank replied:
That is true. The [closest to a reasonable fallback](https://www.gnu.org/software/findutils/manual/html_node/find_html/Systems-without-O_005fNOFOLLOW.html) seems to be equivalent to "check if_symlink, open file, check again", which is still susceptible to timing attacks, just less so. I can do that (and try to use `O_NOFOLLOW` on platforms that support it).

estebank commented:
I've reached a bit of a wall, as I can't use the `md-5` crate in `panicking.rs` (I'm trying to figure out what the issue is, but I guess it is not `no_std` compatible so it can't be referred to from `std`).

the8472 commented:
> > It would have to be a cryptographic hash, but I think that should be enough to close the attack vector.
> 
> Rustc uses MD5 for source hashing because not all DWARF consumers support anything better. And the better thing is SHA1 I believe. And even if it was cryptographically secure, it would still leak if the file on the disk exactly matches the source file.

MD5 and SHA1 should still be fine for this use-case since targeting a specific file with a fixed hash doesn't just require a collision attack but a preimage attack. The algorithms were retired due to collision attacks.

estebank committed:
Increase the size of `Location` with `length` and `hash`

estebank committed:
Add a CRC-32b hash to `SourceFile` and use it for panic snippets

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b92b30951f9fcdb454056538f6df76bcfec438db with merge 2962db43ba3ca77b66ef8717e11faa66f70c362d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b92b30951f9fcdb454056538f6df76bcfec438db","merge_sha":"2962db43ba3ca77b66ef8717e11faa66f70c362d"} -->

estebank commented:
> That's cool! Maybe it would be possible to try using debug info if available to make it more reliable.

Doing so would either not be able to provide as much context or unacceptably increase the size of the binary. I feel like this is mostly useful for the case "I'm developing and running `cargo run` and want to get the same experience for runtime errors I get from the compiler" and nothing more.

> I've reached a bit of a wall, as I can't use the `md-5` crate in `panicking.rs`

Ended up rolling up a crc32 hash myself. I'm now checking what the perf implications of growing `Location` by 64 bytes is.

estebank committed:
Use `is_symlink` so that the check works in all platforms

estebank committed:
Make panic snippet opt-in

`RUST_BACKTRACE=snippet` print only the snippet when possible.
`RUST_BACKTRACE=short,snippet` print the short backtrace and the snippet
when possible.
`RUST_BACKTRACE=full,snippet` print the full backtrace and the snippet
when possible.

estebank committed:
Fix `caller_location` in cranelift

estebank committed:
Tweak what RUST_BACKTRACE accepts

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics/caller_location.rs:
```diff
@@ -107,22 +111,28 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     }
 
-    pub(crate) fn location_triple_for_span(&self, span: Span) -> (Symbol, u32, u32) {
+    pub(crate) fn location_triple_for_span(&self, span: Span) -> (Symbol, u32, u32, u32, u32) {
```
It's no longer a triple... ;)

RalfJung commented on library/std/src/panic.rs:
```diff
@@ -304,6 +316,10 @@ pub fn get_backtrace_style() -> Option<BacktraceStyle> {
                 BacktraceStyle::Full
+            } else if &x == "snippet,full" || &x == "full,snippet" {
+                BacktraceStyle::FullSnippet
+            } else if x.as_os_str().to_str().map_or(false, |s| s.contains(&"snippet")) {
```
This seems very permissive, e.g. it will also fire on `RUST_BACKTRACE=nosnippets`

estebank replied:
I can look for `,snippet` and `snippet,` so that anything that works today for the short backtrace (like "RUST_BACKTRACE=banana") will do what was expected when adding `,snippet`.

estebank replied:
Did the above

estebank committed:
Fix bootstrapping issue

estebank committed:
review comment

estebank committed:
Be more explicit about `RUST_BACKTRACE=snippet`

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f8bb335249042e4949460568eb5f29bb5d9bbd2d with merge 28a23111eea858163483aba013169a190987e657...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f8bb335249042e4949460568eb5f29bb5d9bbd2d","merge_sha":"28a23111eea858163483aba013169a190987e657"} -->

estebank committed:
Cleanup

estebank committed:
Follow up bootstrap fix

eddyb commented on compiler/rustc_const_eval/src/interpret/intrinsics/caller_location.rs:
```diff
@@ -107,22 +111,30 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             .expect("writing to memory we just allocated cannot fail");
         self.write_scalar(col, &self.mplace_field(&location, 2).unwrap().into())
             .expect("writing to memory we just allocated cannot fail");
+        if cfg!(not(bootstrap)) {
```
This is not a bootstrap issue I think. Tests just contain copies of the definition, e.g.:
https://github.com/rust-lang/rust/blob/f722b24eb9ea35a7cc187a1cb5c50d4d324f5855/tests/codegen-units/item-collection/implicit-panic-call.rs#L12-L17

So the fix is to not have this check, but instead update all 10-15 of so copies of that definition.

Or at most, a `loc_layout.fields.count() > 3` comparison here (but not `cfg!(bootstrap)`).

estebank replied:
Yeah, @eddyb just made me realize that the problem was *the test*

marcbowes commented:
I think there are a couple of discussion points here, thought I'd give my input:

1. **What is the default output?** Currently, we print linenumber, and bounds/index information. At a minimum, including column information would be super helpful.
2. **Do we use line/colnumbers to print snippets? If so, under what conditions?** It's really developer friendly to print these out, but there are some security risks. I personally absolutely do not want source code printed in production under any reasons, no matter what heuristics are used. Having an env-var to opt-in seems good to me. Printing "set XX to YY" if that var is not set (like we currently do with backtrace) seems like an easy way to educate developers.
3. **The source file might have changed.** Using checksums seems reasonable, provided the implementation doesn't impact compile times. e.g. if every `unwrap()` generated a checksum and the implementation was slow, I could imagine there could be some degenerative cases (codegen that makes large files full of unwraps based).

bjorn3 commented:
>  personally absolutely do not want source code printed in production under any reasons, no matter what heuristics are used.

In production no source files would be available, right? In that case no snippets will be printed. Source locations have been available since before 1.0 anyway and do respect debuginfo path remapping.

> e.g. if every unwrap() generated a checksum and the implementation was slow, I could imagine there could be some degenerative cases (codegen that makes large files full of unwraps based).

The checksum is for the entire file and calculated only once. We already do another hash of source files when loading them into the compiler's source map anyway. Adding a crc32 doesn't regress performance much.

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit a8526331aa0195d3c549b733803664292bbbde89 with merge 19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a8526331aa0195d3c549b733803664292bbbde89","merge_sha":"19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4208504473/jobs/7306084767)
Build commit: 19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4 (`19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4208504473/jobs/7306084767"},"merge_sha":"19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4"} -->

rust-timer commented:
Queued 19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4 with parent 9aa5c24b7d763fb98d998819571128ff2eb8a3ca, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9aa5c24b7d763fb98d998819571128ff2eb8a3ca&end=19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4).
<!-- rust-timer: temporary -->

wesleywiser reviewed (commented):
Visited during T-compiler triage meeting. Compiler changes look good to me. I think this is really just waiting on T-libs to decide if they have concerns about the runtime behavior or implementation.

m-ou-se commented:
A bunch of thoughts:

I'm not sure it's a good idea to include this code by default in all Rust binaries. I don't think a Rust binary running in production (e.g. on a user's computer) should open files that should only exist on the developer's computer. As an app developer I'm not sure I'd want to ship this to my users.

Formatting panics like a rustc diagnostic (with `-->` and `|` and the line numbers etc.) looks nice, but many people consume their compiler diagnostics not from the terminal, but from an IDE, in which case rustc gives the diagnostics in json rather than with this ascii art. If the goal is to make these as similar as possible to rustc diagnostics, perhaps they should also be reported in the same way.

Perhaps this should be a feature of cargo, such that it is `cargo run` that will format panics in a way that is useful while developing. Maybe something like `RUST_PANIC_FORMAT=json` for output that cargo can parse.

Alternatively, `RUST_ON_PANIC=some-command-to-run` (or `RUST_PANIC_HANDLER=libmypanichandler.so`?) could be an option to make it pluggable, although that will probably also add complexity.

As for the span and file hash information.. I've seen quite a few people asking us to include *less* source information in the resulting binary, mostly from those who ship compiled binaries to their users. So adding even more information seems like a step in the wrong direction. Would it be possible to somehow move this information into the debug information/section of the binaries, in a way that `strip` or similar would strip it from the binary?

> > e.g. if every unwrap() generated a checksum and the implementation was slow, I could imagine there could be some degenerative cases (codegen that makes large files full of unwraps based).
> 
> The checksum is for the entire file and calculated only once. We already do another hash of source files when loading them into the compiler's source map anyway. Adding a crc32 doesn't regress performance much.

You're talking about rustc itself, but I think the comment was about panicking .unwrap()s at runtime, which would each invoke the panic handler, which would read the file again and calculate the hash again at every panic. (E.g. in a server where individual threads may panic while the process continues.)

m-ou-se commented:
The [performance regressions](https://perf.rust-lang.org/compare.html?start=9aa5c24b7d763fb98d998819571128ff2eb8a3ca&end=19df2a290b0ee5caaf8a3216b05f3358c7b7fbc4&stat=instructions:u) look quite significant by the way.

m-ou-se commented on library/std/src/panicking.rs:
```diff
@@ -255,18 +258,116 @@ fn default_hook(info: &PanicInfo<'_>) {
+        let mut print_snippet = || -> Result<(), crate::io::Error> {
+            let filename = location.file();
+            let path = PathBuf::from(filename);
+            // We check if `path` is a symlink to avoid people creating a symlink for a file they
+            // don't have permissions, but that the running binary does, to use it as an
+            // exfiltration vector.
+            let mut path_is_symlink = false;
+            let mut path_check = path.clone();
+            loop {
+                if path_check.is_symlink() {
```
This check can be circumvented with some good timing.

And it won't work if any of the parent directories is a symlink but not the final component. E.g. `/src/symlink/main.rs`.

pnkfelix commented:
I think the feedback from @m-ou-se and @marcbowes should be taken quite seriously here, and that if you combine their feedback together, you get a solution that sounds pretty reasonable to my naive ears:

1. Add the environment variable opt-in for the rustc binary
2. Have {`cargo run`, `cargo test`, etc}  turn that opt-in on by default.

After all, the vast majority of the developers we're seeking to help with this mechanism will be building their code with `cargo` anyway, right? So the extra-step of setting the environment variable will only be suffered by poor souls like me who continue to call `rustc` by hand.

JohnCSimon commented:
@estebank 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!

estebank commented:
I haven't been able to rework this PR to not introduce perfomance regression. Closing.
